### DISCUSSION TITLE
Account Permissions

### DISCUSSION THREAD

**Discussion by ecton** [khonsulabs/cosmicverge#6] [Ideas]
The approach I'm considering for a first pass:

* Accounts have a new flag, superuser. This grants all permissions period.
* Permission Groups will be how all other permissions are assigned.
  * Each permission group has a list of permission grants, which contain a `service` and a `permission`
  * Services would be defined as Rust enum
  * Permissions can be different for each Service, so a plain String type will identify the individual permission
  * All permissions are denied by default
  * A permission grant can enable all permissions for a Service, ie a wildcard match
* Accounts can be assigned one or more permission groups.
* Upon authenticating on the API, the API server will load the permissions granted for the user and send them along to the client so that the client can hide/show the appropriate interfaces after checking permissions.
* Both server and client will use the same API and same structure to query permissions.
* While we are using a String type for permissions, we should use named const static strings.

If we could agree on a list of Permissions that we could list as an Enum, that might be preferrable, but as I was brainstorming types of things I might want to enable permissions for, I felt like a string makes the most sense. For example, Bans are things that happen only for an Account that I could see being split into two permissions: temporary and permanent. Those types of ban permissions wouldn't ever apply to anything else, which is leading me to using a generic string-based approach.

**Comment by daxpedda** [comment]
> Those types of ban permissions wouldn't ever apply to anything else, which is leading me to using a generic string-based approach.

Can you elaborate on that? What do you mean specifically with "anything else", aren't permissions only for accounts? How does that relate to not being able to use `enum`s?

**Comment by ecton** [comment]
Yes, absolutely. I focused on providing a reason to not just have "Read, Write, Delete" style permissions for each section without talking about the other sections. First the use case: I don't think I'll be writing the content for the entire game by myself. I hope to have people such as artists, writers, and community managers that could do things like:

* Add new lore entries, edit existing lore, etc
* Add new assets, view existing assets
* Update loot tables (if that concept exists)

The idea of how to group various areas of permissions was using services -- so we might have services like: Account, Assets, Universe (Planet and solar system management), Ships, Items. Those are ones I am pretty confident we'll want to have, but I could also rattle off more ideas for content relating to missions, npcs, and more.

For these extra contributors to the game, some we'll want to grant just read access to some areas. For example, a copy-editor type role or a localizer role doesn't need permissions to upload new Assets, but they might need to be able to read from the Asset service.

I recognize we could have one large permission enum that has all possible individual permission grants. The system I proposed attempts to make it easy to grant Assets.* for ease-of-use, which I couldn't figure out a good way to do with a giant enum in a generic way. The only use case is granting "admin"-like capabilities over an area, without granting superuser, and not needing to remember to update those groups with any new permissions for an area if we add them later.
